<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Running the pipeline &#8212; amplicon_sequencing_pipeline_doc  documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pipeline output files and directories" href="output.html" />
    <link rel="prev" title="Preparing your dataset for processing" href="preparing-dataset.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="running-the-pipeline">
<h1>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="processing-your-data">
<h2>Processing your data<a class="headerlink" href="#processing-your-data" title="Permalink to this headline">¶</a></h2>
<p>Once you have placed all relevant files and folders, including the
summary file, into a single folder, you can call the script Master.py
with the path to this folder as input. Master.py will parse the summary
file and launch the appropriate scripts to process your request. Suppose
the path to the folder containing the dataset is
<code class="docutils literal"><span class="pre">/home/ubuntu/dataset_folder</span></code>, and contains a summary file where you
specified the dataset ID as ’myDataset’, you would run the following
command from anywhere:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">~/</span><span class="n">scripts</span><span class="o">/</span><span class="n">Master</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">ubuntu</span><span class="o">/</span><span class="n">dataset_folder</span>
</pre></div>
</div>
<p>If you prefer to run the program in the background, you can add an <code class="docutils literal"><span class="pre">&amp;</span></code> to
the end of the command. Alternatively, you can use your favorite
terminal multi-plexer (e.g. <code class="docutils literal"><span class="pre">screen</span></code> or <code class="docutils literal"><span class="pre">tmux</span></code>) to run the code while
continuing your work. Using <code class="docutils literal"><span class="pre">nohup</span></code> ensures that any hangups do not
interrupt the processing:</p>
<div class="code bash highlight-default"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">python</span> <span class="o">~/</span><span class="n">scripts</span><span class="o">/</span><span class="n">Master</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">ubuntu</span><span class="o">/</span><span class="n">dataset_folder</span>
</pre></div>
</div>
<p>Processing happens in a folder within the proc folder, with path
<code class="docutils literal"><span class="pre">/home/ubuntu/proc/myDataset_proc_16S</span></code> or <code class="docutils literal"><span class="pre">/home/ubuntu/proc/myDataset_proc_ITS</span></code>,
depending on the amplicon being analyzed. Final results are put in
<code class="docutils literal"><span class="pre">/home/ubuntu/processing_results/</span></code>. Log files are put in the
<code class="docutils literal"><span class="pre">/home/ubuntu/logs</span> <span class="pre">folder</span></code>, with two files created for each dataset:
<code class="docutils literal"><span class="pre">stderr_datasetID.log</span></code> for error and warning messages and
<code class="docutils literal"><span class="pre">stdout_datasetID.log</span></code> for various progress messages. Certain std out and
std err messages also go to the <code class="docutils literal"><span class="pre">stderr_master.log</span></code> and
<code class="docutils literal"><span class="pre">stdout_master.log</span></code> in the directory from which you called the
<code class="docutils literal"><span class="pre">python</span> <span class="pre">Master.py</span></code> command. Make sure to check both of these places
when troubleshooting!</p>
<p><strong>Please remove your files from the ~/proc folder once you have checked
your processing results!</strong> The ~/proc folder can easily fill up - it
contains the original raw files are and additional copies corresponding
to each processing step!</p>
</div>
<div class="section" id="underlying-function-calls">
<h2>Underlying function calls<a class="headerlink" href="#underlying-function-calls" title="Permalink to this headline">¶</a></h2>
<p>The following table describes the underlying scripts and functions used
in each step of 16S and ITS data processing. Python scripts are
passed-down Alm lab scripts (for the most part).</p>
<p>All code is in <code class="docutils literal"><span class="pre">/home/ubuntu/scripts</span></code> and we encourage users to take a
look!</p>
<table border="1" class="docutils">
<colgroup>
<col width="37%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head"><strong>Processing</strong></th>
<th class="head"><strong>Function/script call</strong></th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Merging</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-fastq_mergepairs</span></code></td>
</tr>
<tr class="row-odd"><td>De-multiplexing</td>
<td><code class="docutils literal"><span class="pre">2.split_by_barcodes.py</span></code></td>
</tr>
<tr class="row-even"><td>Primer trimming</td>
<td><code class="docutils literal"><span class="pre">1.remove_primers.py</span></code></td>
</tr>
<tr class="row-odd"><td>Quality filtering (truncation)</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-fastq_filter</span> <span class="pre">-fastq_truncqual</span></code></td>
</tr>
<tr class="row-even"><td>Quality filtering (expected errors)</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-fastq_filter</span> <span class="pre">-fastq_maxee</span></code></td>
</tr>
<tr class="row-odd"><td>Length trimming (FASTQ)</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-fastq_filter</span> <span class="pre">-fastq_trunclen</span></code></td>
</tr>
<tr class="row-even"><td>Length trimming (FASTA)</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-fastx_truncate</span> <span class="pre">-trunclen</span></code></td>
</tr>
<tr class="row-odd"><td>Dereplication</td>
<td><code class="docutils literal"><span class="pre">3.dereplicate.py</span></code></td>
</tr>
<tr class="row-even"><td>De novo clustering</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-cluster_otus</span> <span class="pre">-otu_radius_pct</span></code></td>
</tr>
<tr class="row-odd"><td>Closed-reference mapping</td>
<td><code class="docutils literal"><span class="pre">usearch8</span> <span class="pre">-usearch_global</span> <span class="pre">-db</span> <span class="pre">GG_database_file</span> <span class="pre">-strand</span> <span class="pre">both</span></code></td>
</tr>
<tr class="row-even"><td>RDP taxonomy assignment</td>
<td><code class="docutils literal"><span class="pre">rdp_classify.py</span></code></td>
</tr>
<tr class="row-odd"><td>Distribution-based clustering</td>
<td><code class="docutils literal"><span class="pre">dbotu.py</span></code> <code class="docutils literal"><span class="pre">call_otus()</span></code></td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Running the pipeline</a><ul>
<li><a class="reference internal" href="#processing-your-data">Processing your data</a></li>
<li><a class="reference internal" href="#underlying-function-calls">Underlying function calls</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="preparing-dataset.html" title="previous chapter">Preparing your dataset for processing</a></li>
      <li>Next: <a href="output.html" title="next chapter">Pipeline output files and directories</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/running.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Thomas Gurry and Claire Duvallet.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
      |
      <a href="_sources/running.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>